File: src/Cassandra.php
<?php

/**
 * Copyright 2015 DataStax, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use Cassandra\Cluster\Builder as ClusterBuilder;
use Cassandra\SSLOptions\Builder as SSLOptionsBuilder;

/**
 * The main entry point to the PHP Driver for Apache Cassandra
 *
 * Use Cassandra::cluster() to build a cluster instance.
 * Use Cassandra::ssl() to build SSL options instance.
 */
final class Cassandra
{
    const CONSISTENCY_ANY          = 0;
    const CONSISTENCY_ONE          = 1;
    const CONSISTENCY_TWO          = 2;
    const CONSISTENCY_THREE        = 3;
    const CONSISTENCY_QUORUM       = 4;
    const CONSISTENCY_ALL          = 5;
    const CONSISTENCY_LOCAL_QUORUM = 6;
    const CONSISTENCY_EACH_QUORUM  = 7;
    const CONSISTENCY_SERIAL       = 8;
    const CONSISTENCY_LOCAL_SERIAL = 9;
    const CONSISTENCY_LOCAL_ONE    = 10;

    const VERIFY_NONE          = 0;
    const VERIFY_PEER_CERT     = 1;
    const VERIFY_PEER_IDENTITY = 2;

    const BATCH_LOGGED   = 0;
    const BATCH_UNLOGGED = 1;
    const BATCH_COUNTER  = 2;

    const LOG_DISABLED = 0;
    const LOG_CRITICAL = 1;
    const LOG_ERROR    = 2;
    const LOG_WARN     = 3;
    const LOG_INFO     = 4;
    const LOG_DEBUG    = 5;
    const LOG_TRACE    = 6;

    const TYPE_TEXT      = "text";
    const TYPE_ASCII     = "ascii";
    const TYPE_VARCHAR   = "varchar";
    const TYPE_BIGINT    = "bigint";
    const TYPE_BLOB      = "blob";
    const TYPE_BOOLEAN   = "boolean";
    const TYPE_COUNTER   = "counter";
    const TYPE_DECIMAL   = "decimal";
    const TYPE_DOUBLE    = "double";
    const TYPE_FLOAT     = "float";
    const TYPE_INT       = "int";
    const TYPE_TIMESTAMP = "timestamp";
    const TYPE_UUID      = "uuid";
    const TYPE_VARINT    = "varint";
    const TYPE_TIMEUUID  = "timeuuid";
    const TYPE_INET      = "inet";

    const LOAD_BALANCING_ROUND_ROBIN          = 0;
    const LOAD_BALANCING_DC_AWARE_ROUND_ROBIN = 1;

    const VERSION = '1.0.0-alpha';

    /**
     * Returns a Cluster Builder
     *
     * @return Cassandra\Cluster\Builder a Cluster Builder instance
     */
    public static function cluster()
    {
        return new ClusterBuilder();
    }

    /**
     * Returns SSL Options Builder
     *
     * @return Cassanrda\SSLOptions\Builder a SSLOptions Builder instance
     */
    public static function ssl()
    {
        return new SSLOptionsBuilder();
    }

    /**
     * Sets log severity
     *
     * @param  int  $level  log severity, must be one of Cassandra::LOG_* constants.
     * @return void
     */
    public static function setLogLevel($level)
    {
        cassandra_set_log_level($level);
    }
}

File: ext/src/Cassandra.c
#include "php_cassandra.h"

zend_class_entry *cassandra_ce = NULL;

ZEND_EXTERN_MODULE_GLOBALS(cassandra)

PHP_METHOD(Cassandra, cluster)
{
  object_init_ex(return_value, cassandra_cluster_builder_ce);
}

PHP_METHOD(Cassandra, ssl)
{
  object_init_ex(return_value, cassandra_ssl_builder_ce);
}

PHP_METHOD(Cassandra, setLogLevel)
{
  zval* level;

  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z", &level) == FAILURE) {
    return;
  }

  if (Z_TYPE_P(level) != IS_LONG) {
    INVALID_ARGUMENT(level, "one of Cassandra::LOG_*");
  }

  switch (Z_LVAL_P(level)) {
  case CASS_LOG_DISABLED:
  case CASS_LOG_CRITICAL:
  case CASS_LOG_ERROR:
  case CASS_LOG_WARN:
  case CASS_LOG_INFO:
  case CASS_LOG_DEBUG:
  case CASS_LOG_TRACE:
    break;
  default:
    INVALID_ARGUMENT(level, "one of Cassandra::LOG_*");
  }

  CASSANDRA_G(log_level) = Z_LVAL_P(level);
}

ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_level, 0, ZEND_RETURN_VALUE, 0)
  ZEND_ARG_INFO(0, level)
ZEND_END_ARG_INFO()

static zend_function_entry Cassandra_methods[] = {
  PHP_ME(Cassandra, cluster,     arginfo_none,  ZEND_ACC_PUBLIC|ZEND_ACC_STATIC)
  PHP_ME(Cassandra, ssl,         arginfo_none,  ZEND_ACC_PUBLIC|ZEND_ACC_STATIC)
  PHP_ME(Cassandra, setLogLevel, arginfo_level, ZEND_ACC_PUBLIC|ZEND_ACC_STATIC)
  PHP_FE_END
};

void cassandra_define_Cassandra(TSRMLS_D)
{
  zend_class_entry ce;

  INIT_CLASS_ENTRY(ce, "Cassandra", Cassandra_methods);
  cassandra_ce = zend_register_internal_class(&ce TSRMLS_CC);

  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_ANY"),          CASS_CONSISTENCY_ANY TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_ONE"),          CASS_CONSISTENCY_ONE TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_TWO"),          CASS_CONSISTENCY_TWO TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_THREE"),        CASS_CONSISTENCY_THREE TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_QUORUM"),       CASS_CONSISTENCY_QUORUM TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_ALL"),          CASS_CONSISTENCY_ALL TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_LOCAL_QUORUM"), CASS_CONSISTENCY_LOCAL_QUORUM TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_EACH_QUORUM"),  CASS_CONSISTENCY_EACH_QUORUM TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_SERIAL"),       CASS_CONSISTENCY_SERIAL TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_LOCAL_SERIAL"), CASS_CONSISTENCY_LOCAL_SERIAL TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("CONSISTENCY_LOCAL_ONE"),    CASS_CONSISTENCY_LOCAL_ONE TSRMLS_CC);

  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("VERIFY_NONE"),          CASS_SSL_VERIFY_NONE TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("VERIFY_PEER_CERT"),     CASS_SSL_VERIFY_PEER_CERT TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("VERIFY_PEER_IDENTITY"), CASS_SSL_VERIFY_PEER_IDENTITY TSRMLS_CC);

  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("BATCH_LOGGED"),   CASS_BATCH_TYPE_LOGGED TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("BATCH_UNLOGGED"), CASS_BATCH_TYPE_UNLOGGED TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("BATCH_COUNTER"),  CASS_BATCH_TYPE_COUNTER TSRMLS_CC);

  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOG_DISABLED"), CASS_LOG_DISABLED TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOG_CRITICAL"), CASS_LOG_CRITICAL TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOG_ERROR"),    CASS_LOG_ERROR TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOG_WARN"),     CASS_LOG_WARN TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOG_INFO"),     CASS_LOG_INFO TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOG_DEBUG"),    CASS_LOG_DEBUG TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOG_TRACE"),    CASS_LOG_TRACE TSRMLS_CC);

  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_TEXT"),      "text" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_ASCII"),     "ascii" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_VARCHAR"),   "varchar" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_BIGINT"),    "bigint" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_BLOB"),      "blob" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_BOOLEAN"),   "boolean" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_COUNTER"),   "counter" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_DECIMAL"),   "decimal" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_DOUBLE"),    "double" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_FLOAT"),     "float" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_INT"),       "int" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_TIMESTAMP"), "timestamp" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_UUID"),      "uuid" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_VARINT"),    "varint" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_TIMEUUID"),  "timeuuid" TSRMLS_CC);
  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("TYPE_INET"),      "inet" TSRMLS_CC);

  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOAD_BALANCING_ROUND_ROBIN"),          0 TSRMLS_CC);
  zend_declare_class_constant_long(cassandra_ce, ZEND_STRL("LOAD_BALANCING_DC_AWARE_ROUND_ROBIN"), 1 TSRMLS_CC);

  zend_declare_class_constant_string(cassandra_ce, ZEND_STRL("VERSION"), "1.0.0-beta" TSRMLS_CC);
}
